### PR TITLE
Replace direct usage of ClassLoader with ResourcesLoader

### PR DESCRIPTION
**Pull request by riccardobl**
To enhance compatibility with JVMs that deviate from standard norms and may lack complete or partial support for classloaders (such as teavm), this pull request introduces a new encapsulating class called ResourcesLoader. Functioning analogously to BufferAllocators, this addition empowers developers to specify a personalized implementation. The default implementation, named ResourcesLoaderJImpl, operates by interfacing with the classloader.

Throughout the core, all instances where classloaders were used have been substituted with the utilization of ResourcesLoader, excluding specific parts in GL debug classes and jme3-networking which involve proxies and might need more careful consideration.

The methods within AssetManager that depend on ClassLoader usage have been deprecated and could potentially be marked for removal, as they are not utilized by the engine itself.

### CODE REVIEW

File: jme3-core/src/main/java/com/jme3/util/res/ResourcesLoaderImpl.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.jme3.util.res;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Enumeration;
+
+public interface ResourcesLoaderImpl {
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/util/res/ResourcesLoaderImpl.java:39]: I think it is common to use the "Impl" suffix name on the classes that implement the interface and is not used for naming the interface itself, but I might be wrong!

Comment (pspeed42) [reply jme3-core/src/main/java/com/jme3/util/res/ResourcesLoaderImpl.java:39]: Yes, *Impl = concrete class

**Comment by Ali-RS** [comment]
Curious to know how a non-JVM-based resource loader would look?

File: jme3-core/src/main/java/com/jme3/util/res/ResourcesLoader.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    
+
+   public static Enumeration<URL> getResources(String path ) throws IOException {
+        return getImpl().getResources(path,null);        
+    }
+
+
+   public static Enumeration<URL> getResources(String path,Class<?> clazz) throws IOException {
+        return getImpl().getResources(path,clazz);        
+    }
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/util/res/ResourcesLoader.java:100]: Please fix the formatting issues in the above methods. (add required whitespace and remove extras)

**Comment by riccardobl** [comment]
> Do you have an implementation of ResourceLoader that is compatible with Teavm?
Yes, the resource loader for teavm loads from the site root

```java
package com.jme3.web.filesystem;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Enumeration;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.teavm.jso.browser.Window;

import com.jme3.util.res.ResourcesLoaderImpl;

public class WebResourceLoaderImpl implements ResourcesLoaderImpl {
    private static final Logger logger = Logger.getLogger(WebResourceLoaderImpl.class.getName());

    private String getFullPath(Class<?> clazz, String path) {
        String resourcePath = path;
        if (clazz != null) {
            String className = clazz.getName();
            String classPath = className.replace('.', '/') + ".class";
            classPath = classPath.substring(0, classPath.lastIndexOf('/'));
            resourcePath = classPath + "/" + path;
        }
        try {
            URL baseURL = new URL(Window.current().getLocation().getFullURL());
            resourcePath = new URL(baseURL, resourcePath).toString();
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
        return resourcePath;
    }

    @Override
    public URL getResource(String path, Class<?> clazz) {
        path = getFullPath(clazz, path);
        logger.log(Level.FINE, "Fetch resource {0}", path);
        try {
            URL url = new URL(path);
            return url;
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public InputStream getResourceAsStream(String path, Class<?> clazz) {
        path = getFullPath(clazz, path);
        logger.log(Level.FINE, "Fetch resource as stream {0}", path);
        try {
            URL url = new URL(path);
            InputStream is = url.openStream();
            return is;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public Enumeration<URL> getResources(String path, Class<?> clazz) throws IOException {
        List<URL> urls = new ArrayList<URL>();
        try {
            String indexPath = "/index.dat";
            logger.log(Level.FINE, "Fetch resources index {0}", indexPath);
            URL resourcesIndex = getResource(indexPath, null);
            ByteArrayOutputStream resourcesBo = new ByteArrayOutputStream();
            byte[] buffer = new byte[1024];
            int read = 0;
            InputStream is = resourcesIndex.openStream();
            while ((read = is.read(buffer)) > 0) {
                resourcesBo.write(buffer, 0, read);
            }
            is.close();
            logger.log(Level.FINE, "Listing resources in {0}", path);
            String resources[] = new String(resourcesBo.toByteArray(), Charset.forName("UTF-8")).split("\n");
            for (String resource : resources) {
                resource = resource.trim();
                if (resource.length() > 0 && resource.startsWith(path)) {
                    URL url = getResource(resource, null);
                    urls.add(url);
                }
            }
            logger.log(Level.FINE, "Found {0} resources in {1}", new Object[] { urls.size(), path });
        } catch (Exception e) {
            e.printStackTrace();
        }
        return Collections.enumeration(urls);
    }

}
```

File: jme3-core/src/main/java/com/jme3/util/res/ResourcesLoaderImpl.java (hunk 1 above)
Comment (riccardobl) [reply jme3-core/src/main/java/com/jme3/util/res/ResourcesLoaderImpl.java]: Renamed as you proposed

File: jme3-core/src/main/java/com/jme3/util/res/ResourcesLoader.java (hunk 2 above)
Comment (riccardobl) [reply jme3-core/src/main/java/com/jme3/util/res/ResourcesLoader.java]: Done

File: jme3-core/src/main/java/com/jme3/util/res/DefaultResourceLoader.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ */
+package com.jme3.util.res;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Enumeration;
+
+/**
+ * Default java-like implementation of ResourceLoaderImpl. Loads from classpath.
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/util/res/DefaultResourceLoader.java:40]: ResourceLoaderImpl in the Javadoc should be renamed to ResourceLoader.

File: jme3-core/src/plugins/java/com/jme3/asset/plugins/ClasspathLocator.java [hunk 4]
```diff
@@ -87,20 +89,10 @@ public AssetInfo locate(AssetManager manager, AssetKey key) {
...
-            for (final ClassLoader classLoader : classLoaders) {
-                url = classLoader.getResource(name);
-                if(url != null) {
-                    break;
-                }
-            }
+        if (JmeSystem.isLowPermissions()) {
+            url = Resources.getResource("/" + name, ClasspathLocator.class);    
+        } else {
+            url = Resources.getResource( name);
```
Comment (Ali-RS) [inline jme3-core/src/plugins/java/com/jme3/asset/plugins/ClasspathLocator.java:95]: please remove the trailing space in `Resources.getResource( name); `

File: jme3-core/src/main/java/com/jme3/util/res/ResourcesLoaderImpl.java (hunk 1 above)
Comment (Ali-RS) [reply jme3-core/src/main/java/com/jme3/util/res/ResourcesLoaderImpl.java]: Thanks

File: jme3-core/src/main/java/com/jme3/util/res/ResourcesLoader.java (hunk 2 above)
Comment (Ali-RS) [reply jme3-core/src/main/java/com/jme3/util/res/ResourcesLoader.java]: Thanks

File: jme3-core/src/main/java/com/jme3/util/res/Resources.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            try {
+                impl = (ResourceLoader) Class.forName(DEFAULT_IMPL).getDeclaredConstructor().newInstance();
+            } catch (final Throwable e1) {
+                throw new RuntimeException("Unable to access default resources loader implementation", e1);
+            }
+        }
+        return impl;
+    }
+
+    public static void setImpl(ResourceLoader impl) {
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/util/res/Resources.java:68]: Can you please rename this method to `setResourceLoader()`?

File: jme3-core/src/main/java/com/jme3/util/res/Resources.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+ * classpath.
+ */
+public class Resources {
+    public static final String PROPERTY_RESOURCE_LOADER_IMPLEMENTATION = "com.jme3.ResourceLoaderImplementation";
+    private static final String DEFAULT_IMPL = "com.jme3.util.res.DefaultResourceLoader";
+
+    private static final Logger LOGGER = Logger.getLogger(Resources.class.getName());
+    private static ResourceLoader impl = null;
+
+    private static ResourceLoader getImpl() {
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/util/res/Resources.java:52]: Can you please rename this method to `getResourceLoader()`?

File: jme3-core/src/main/java/com/jme3/util/res/ResourceLoader.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Enumeration;
+
+public interface ResourceLoader {
+    /**
+     * Finds the resource with the given name relative to the given parent class
+     * or to the root of the parent is null.
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/util/res/ResourceLoader.java:42]: >or to the root of the parent is null.

Typo? (`of` -> `if`)

File: jme3-core/src/main/java/com/jme3/util/res/ResourceLoader.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+     *            The resource name
+     * @param parent
+     *            Optional parent class
+     * @return The resource URL or null if not found
+     */
+    public URL getResource(String path, Class<?> parent);
+
+    /**
+     * Finds the resource with the given name relative to the given parent class
+     * or to the root of the parent is null.
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/util/res/ResourceLoader.java:54]: Same here

File: jme3-core/src/main/java/com/jme3/util/res/ResourceLoader.java (hunk 7 above)
Comment (riccardobl) [reply jme3-core/src/main/java/com/jme3/util/res/ResourceLoader.java]: Thanks, fixed

**Comment by MeFisto94** [comment]
When reviewing, I've asked myself what the actual problem/purpose here is and to me I could come up with 2-3 actual different use-cases:
1. We seem to have quite some code that loads ominous .cfg files and by doing so, it doesn't even leverage the asset manager but uses class.getResourceAsStream(). This in itself may be questionable, because you have a config file, but can maybe hardly overwrite it? And if you're supposed to override it, why can't we use the AssetManager for such, perspectively?
2. There are some cases that have a legitimate class loader, especially around the `Savable` interface. In order to properly deserialize, we need class instances and it looks like the class loader just tries to locate and instantiate those (instead of manually registering implementations or `Class<?>`s.
3. Then there's the ClassLoader AssetLocator, probably the default for DesktopAssetManager, but do we really need to replace it with a ResourceLoader, if we could just swap in another AssetLocator?

So basically I just have the gut feeling that we could solve each of those even better, given good design/refactorings.

**Comment by riccardobl** [comment]
> We seem to have quite some code that loads ominous .cfg files and by doing so, it doesn't even leverage the asset manager but uses class.getResourceAsStream(). This in itself may be questionable, because you have a config file, but can maybe hardly overwrite it? And if you're supposed to override it, why can't we use the AssetManager for such, perspectively?

These files are used to initialize the assetManager , so it doesn't exist yet at that point. Also it is a good thing they don't use the asset manage features, because they aren't assets, they don't need caching or to be overwritten by things in the asset path.


> 2\. There are some cases that have a legitimate class loader, especially around the `Savable` interface. In order to properly deserialize, we need class instances and it looks like the class loader just tries to locate and instantiate those (instead of manually registering implementations or `Class<?>`s.

This change is only related to resources in the classpath, not classes.





> Then there's the ClassLoader AssetLocator, probably the default for DesktopAssetManager, but do we really need to replace it with a ResourceLoader, if we could just swap in another AssetLocator?

If the problem was only with the asset locator, i would have written a new one of my teavm implementation, but this affect various parts of the engine outside of the asset locator and the side effect of this patch is that the default classpath locator becomes automatically compatible.




> So basically I just have the gut feeling that we could solve each of those even better, given good design/refactorings.

Maybe, but redesign all this would take way too much time vs this simple patch that doesn't break anything.

File: jme3-core/src/main/java/com/jme3/util/res/Resources.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * This class is used to load resources from the default location usually the
+ * classpath.
+ */
+public class Resources {
+    public static final String PROPERTY_RESOURCE_LOADER_IMPLEMENTATION = "com.jme3.ResourceLoaderImplementation";
+    private static final String DEFAULT_IMPL = "com.jme3.util.res.DefaultResourceLoader";
```
[the remark this answers is not in the document]
Comment (riccardobl) [reply jme3-core/src/main/java/com/jme3/util/res/Resources.java]: That's done on purpose so that "com.jme3.util.res.DefaultResourceLoader"  doesn't need to be included in the build.
